### DISCUSSION TITLE
Phase 6.3 Show & Tell — EWC penalty math, parameter importance maps, and catastrophic forgetting prevention

### DISCUSSION THREAD

**Discussion by web3guru888** [web3guru888/asi-build#250] [Show and tell]
## Phase 6.3 Show & Tell: EWC Penalty Math, Parameter Importance Maps, and Catastrophic Forgetting Prevention

This Show & Tell digs into the mathematics behind Elastic Weight Consolidation, how to visualize Fisher information as a "parameter importance map," and what catastrophic forgetting actually looks like in a spike-timing system.

---

## 1. The EWC Loss Function — Derivation

Standard EWC adds a quadratic penalty to the loss to protect important parameters from changing too much:

```
L_EWC = L_task(θ) + λ/2 · Σᵢ Fᵢ · (θᵢ − θᵢ*)²
```

Where:
- `L_task(θ)` — current task loss (STDP weight mismatch)
- `Fᵢ` — diagonal Fisher information for parameter `i` (estimated empirically)
- `θᵢ*` — anchor weights from the previous SLEEP_PHASE consolidation
- `λ` — EWC regularization strength (trade-off hyperparameter)

The **gradient** of the penalty w.r.t. `θᵢ` is:

```
∂L_EWC/∂θᵢ = λ · Fᵢ · (θᵢ − θᵢ*)
```

In code (Phase 6.3 implementation):
```python
grad = 2.0 * ewc_lambda * fisher_diag * (weights - anchor_weights)
```

---

## 2. Fisher Information — What It Measures

The Fisher information matrix **I(θ)** measures how sensitive the model likelihood is to each parameter. High Fisher = parameter critically affects outputs = protect it strongly.

**Diagonal approximation** (used in ASI-Build for tractability):
```
F̂ᵢ = E[(∂ log p(y|x,θ)/∂θᵢ)²]
```

In practice (empirical Fisher from weight deltas):
```python
fisher_diag[i] = mean(delta_w[i]**2 over recent K episodes)
```

**Interpretation table:**

| Fisher value | Meaning | EWC effect |
|---|---|---|
| `Fᵢ ≈ 0` | Parameter barely used | Allow free movement |
| `Fᵢ = 0.1` | Mildly important | Light constraint |
| `Fᵢ = 1.0` | Important | Moderate constraint |
| `Fᵢ = 10.0` | Critical | Strong constraint |
| `Fᵢ > 100` | Foundational | Near-frozen |

---

## 3. Parameter Importance Map — Visualization

For an N × M weight matrix, importance can be rendered as a heatmap in Grafana using `phase6_ewc_penalty_magnitude` histogram:

```promql
rate(phase6_ewc_penalty_magnitude_bucket{task_id="navigation"}[1m])
```

This shows penalty intensity over time — spikes indicate the learner trying to move away from consolidated parameters. High sustained penalty = healthy constraint; zero = no Fisher data yet (first episode).

---

## 4. Catastrophic Forgetting in a Spike-Timing System

Without EWC, after learning Task B, performance on Task A degrades:

```
Task A accuracy:
Before Task B:  ████████████████████ 95%
After Task B:   ████░░░░░░░░░░░░░░░░ 22%  ← catastrophic
With EWC:       ████████████████░░░░ 81%  ← protected
```

In STDP terms: the synaptic weights for Task A firing patterns get overwritten by Task B LTP/LTD updates. EWC prevents this by penalising changes to synapses that fired reliably during Task A.

---

## 5. Design Decisions in Phase 6.3

| Decision | Choice | Rationale |
|---|---|---|
| Gradient or direct penalty? | Gradient (subtract from delta) | Composable with STDP kernel; no loss function needed |
| Per-update vs. batch penalty | `per_update` default | Real-time constraint; lower memory |
| Clip threshold | `max_penalty_clip=1.0` | Prevents Fisher=1e6 from freezing all learning |
| Task isolation | `task_id` parameter | Multi-agent: each agent has its own Fisher store partition |
| Backward compatibility | `ewc=None` pass-through | Existing STDP users unaffected |

---

## 6. Expected Metrics After Integration

After Phase 6.3 lands, Prometheus should show:

```
phase6_ewc_penalty_magnitude{task_id="navigation"}  → histogram, mean ~0.3 per update
phase6_ewc_penalty_clipped_total                    → rare; spikes = Fisher instability
phase6_weight_delta_before_ewc                      → should exceed after_ewc (penalty subtractive)
phase6_weight_delta_after_ewc                       → effective learning delta after constraint
```

---

## Open Questions

1. **Online Fisher update**: Should `FisherMatrixBase.update()` be called during LEARNING (running estimate) or only at SLEEP_PHASE (episodic snapshot)?
2. **Task ID strategy**: In a multi-agent Rings Network, should `task_id` be agent UUID, episode tag, or semantic task label?
3. **EWC warm-up**: During the very first LEARNING phase (no snapshot yet), EWC returns zero gradient — is this the right default or should there be a prior?
